dali/pipeline/pipeline.cc

```diff
@@ -162,15 +162,30 @@ Pipeline::~Pipeline() {
     OperatorBase *op_ptr = &node.InstantiateOperator();
     auto op_type = graph_.NodeType(node_id);
     switch (op_type) {
-      case OpType::CPU:
-        dynamic_cast<InputOperator<CPUBackend> *>(op_ptr)->BreakWaiting();
-        break;
-      case OpType::MIXED:
-        dynamic_cast<InputOperator<MixedBackend> *>(op_ptr)->BreakWaiting();
-        break;
-      case OpType::GPU:
-        dynamic_cast<InputOperator<GPUBackend> *>(op_ptr)->BreakWaiting();
-        break;
+      case OpType::CPU: {
+        auto *cpu_op_ptr = dynamic_cast<InputOperator<CPUBackend> *>(op_ptr);
+        if (!cpu_op_ptr) {
+          std::cerr << "Failed to break waiting - expected CPU op" << std::endl;
+          std::terminate();
+        }
+        cpu_op_ptr->BreakWaiting();
+      } break;
+      case OpType::MIXED: {
+        auto *mixed_op_ptr = dynamic_cast<InputOperator<MixedBackend> *>(op_ptr);
+        if (!mixed_op_ptr) {
+          std::cerr << "Failed to break waiting - expected MIXED op" << std::endl;
+          std::terminate();
+        }
+        mixed_op_ptr->BreakWaiting();
+      } break;
+      case OpType::GPU: {
+        auto *gpu_op_ptr = dynamic_cast<InputOperator<GPUBackend> *>(op_ptr);
+        if (!gpu_op_ptr) {
+          std::cerr << "Failed to break waiting - expected GPU op" << std::endl;
+          std::terminate();
+        }
+        gpu_op_ptr->BreakWaiting();
+      } break;
```

Comment:
Not entirely sure about it, but maybe it is better to fail in more controlled manner here.